katdal/datasources.py

```diff
@@ -53,24 +151,73 @@ def name(self):
 
 class TelstateDataSource(DataSource):
     """A data source based on :class:`katsdptelstate.TelescopeState`."""
-    def __init__(self, telstate, source_name='telstate'):
+    def __init__(self, telstate, capture_block_id=None, stream_name=None,
+                 source_name='telstate'):
+        # Create telstate view based on capture block ID and/or stream name
+        if stream_name:
+            telstate = telstate.view(stream_name)
+        if capture_block_id:
+            telstate = telstate.view(capture_block_id)
+        if stream_name and capture_block_id:
+            cb_stream = telstate.SEPARATOR.join((capture_block_id, stream_name))
+            telstate = telstate.view(cb_stream)
         self.telstate = telstate
-        attrs = {}
+        # Collect sensors
         sensors = {}
         for key in telstate.keys():
-            if telstate.is_immutable(key):
-                attrs[key] = telstate[key]
-            else:
+            if not telstate.is_immutable(key):
                 sensors[key] = TelstateSensorData(telstate, key)
-        metadata = AttrsSensors(attrs, sensors, name=source_name)
-        DataSource.__init__(self, metadata, None)
+        metadata = AttrsSensors(telstate, sensors, name=source_name)
+        try:
+            base_name = telstate['chunk_name']
+            chunk_info = telstate['chunk_info']
+        except KeyError:
+            # Metadata without data
+            DataSource.__init__(self, metadata, None)
+        else:
+            # Extract VisFlagsWeights and timestamps from telstate
+            store = RadosChunkStore.from_config(telstate['ceph_conf'],
+                                                telstate['ceph_pool'])
+            ts_name = store.join(base_name, 'timestamps')
+            ts_chunks = chunk_info['timestamps']['chunks']
+            ts_dtype = chunk_info['timestamps']['dtype']
+            timestamps = store.get_dask_array(ts_name, ts_chunks, ts_dtype)
+            timestamps = DaskLazyIndexer(timestamps)
```

Comment:
Will having lazy timestamps be an issue for performance i.e. is katdal going to do a lot of operations on the timestamps array, each of which requires a fresh request to the store? Or is katdal already designed around timestamps being lazy-indexed?

Reply:
This has already changed in the following V4 PR #119, plus this will all go away anyway once the timestamps are synthesised by me.